pdg/src/main.rs

```diff
@@ -50,10 +51,104 @@ impl Display for ToPrint {
     }
 }
 
+pub struct Pdg {
+    pub events: Vec<Event>,
+    pub metadata: Metadata,
+    pub graphs: Graphs,
+}
+
+impl Pdg {
+    pub fn new(metadata_path: &Path, event_log_path: &Path) -> eyre::Result<Self> {
+        let events = read_event_log(event_log_path)?;
+        let metadata = read_metadata(metadata_path)?;
+        let graphs = construct_pdg(&events, &metadata);
+        Ok(Self {
+            events,
+            metadata,
+            graphs,
+        })
+    }
+
+    pub fn repr<'a>(&'a self, to_print: &'a [ToPrint]) -> PdgRepr<'a> {
+        PdgRepr {
+            pdg: self,
+            to_print,
+        }
+    }
+}
+
+pub struct PdgRepr<'a> {
+    pub pdg: &'a Pdg,
+    pub to_print: &'a [ToPrint],
+}
+
+impl Display for PdgRepr<'_> {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        let Self {
+            pdg:
+                Pdg {
+                    events,
+                    metadata,
+                    graphs,
+                },
+            to_print,
+        } = self;
+        let should_print = |e| to_print.contains(&e);
+
+        if should_print(ToPrint::Metadata) {
+            writeln!(f, "{metadata:#?}")?;
+        }
+
+        if should_print(ToPrint::Events) {
+            for event in events.iter() {
+                let mir_loc = metadata.get(event.mir_loc);
+                let kind = &event.kind;
+                writeln!(f, "{mir_loc:?} -> {kind:?}")?;
+            }
+        }
+
+        if should_print(ToPrint::LatestAssignments) {
+            for ((func_hash, local), p) in &graphs.latest_assignment {
+                let func = &metadata.functions[func_hash];
+                writeln!(f, "({func}:{local:?}) => {p:?}")?;
+            }
+        }
+
+        for graph in &graphs.graphs {
+            let needs_write = graph
+                .needs_write_permission()
+                .map(|node_id| node_id.as_usize())
+                .collect::<Vec<_>>();
```

Comment:
This can be moved inside of `if should_print(ToPrint::WritePermissions) { ... }`.

Reply:
(Same with this) I'd like to change this in a separate PR, since this right now is just moving code, not changing anything.